### ISSUE TITLE
Cursor.Last() returns nil for non-empty bucket

### ISSUE THREAD

**Issue by hrissan** [etcd-io/bbolt#95]
Ubuntu amd-64 go 1.9.2

```go
package main

import (
	"fmt"
	"log"
	"time"

	"github.com/coreos/bbolt"
)

var (
	testBucket = []byte("test")
)

func initDB(path string) (*bolt.DB, error) {
	db, err := bolt.Open(path, 0600, &bolt.Options{Timeout: 1 * time.Second})
	if err != nil {
		return nil, err
	}
	err = db.Update(func(tx *bolt.Tx) error {
		err := tx.DeleteBucket(testBucket)
		_, err = tx.CreateBucketIfNotExists(testBucket)
		if err != nil {
			return err
		}
		return nil
	})
	if err != nil {
		db.Close()
		return nil, err
	}

	return db, nil
}

func formatHeight(a int) []byte {
	return []byte(fmt.Sprintf("%08d", a))
}

func delBackwards(cb *bolt.Bucket, index int) error {
	err := cb.Delete(formatHeight(index))
	if err != nil {
		return fmt.Errorf("delete failed: %v", err)
	}
	cc := cb.Cursor()
	k, _ := cc.Last()
	k2, _ := cc.First()
	if k == nil && k2 != nil {
		return fmt.Errorf("last and first are inconsistent at %d for %#v %#v", index, k, k2)
	}
	return nil
}

func test(n int, value string) {
	db, err := initDB("bad_bolt.db")
	if err != nil {
		log.Fatalf("failed to initialize database: %v", err)
	}
	defer db.Close()
	err = db.Update(func(tx *bolt.Tx) error {
		cb := tx.Bucket(testBucket)
		for i := 0; i < n; i += 1 {
			err := cb.Put(formatHeight(i), []byte(value))
			if err != nil {
				return fmt.Errorf("put failed failed: %v", err)
			}
		}
		return nil
	})
	if err != nil {
		log.Fatalf("first tx failed: %v", err)
	}
	err = db.Update(func(tx *bolt.Tx) error {
		cb := tx.Bucket(testBucket)
		for i := n - 1; i >= 0; i -= 1 {
			err := delBackwards(cb, i)
			if err != nil {
				return err
			}
		}
		return nil
	})
	if err != nil {
		log.Printf("second tx failed: %v", err)
	} else {
		log.Printf("%d %v OK", n, value)
	}
}

func main() {
	test(100, "0000000000000000")
	test(100, "00000000000000000")
}
```

**Comment by Hittorp** [comment]
Is it enough to put n.rebalance() at the end of (n *node) del(key []byte) to fix this?

I am not sure about invariants that should provide cursor.Delete(). Is it ok to get an empty leaf page after delete a key?

**Comment by cenkalti** [comment]
I have run the provided program hundreds of time but couldn't reproduce the issue mentioned. It's probably fixed due to this issue is quite old hence closing it. Feel free to reply and re-open if it is still valid.

**Comment by ahrtr** [comment]
FYI. This is a real issue, and was fixed in 1.3.7 in https://github.com/etcd-io/bbolt/pull/341.
